backend/core/utils/tool.py

```diff
@@ -0,0 +1,156 @@
+                raise error
+
+    def cursor(self):
+        if not self._cursor or self._cursor.closed:
+            if not self._connection:
+                self.connect()
+            self._cursor = self._connection.cursor()
+            return self._cursor
+
+    def execute(self, query, vals=None, retry_counter=0):
+        try:
+            if vals:
+                self._cursor.execute(query, vals)
+            else:
+                self._cursor.execute(query)
+            retry_counter = 0
+        except (psycopg2.DatabaseError, psycopg2.OperationalError) as error:
+            if retry_counter >= LIMIT_RETRIES:
+                raise error
+            retry_counter += 1
+            print("got error {}. retrying {}".format(str(error).strip(), retry_counter))
+            time.sleep(1)
+            self.reset()
+            self.execute(query, vals, retry_counter)
+        except (Exception, psycopg2.Error) as error:
+            if not self.readonly:
+                self._connection.rollback()
+            raise error
+        else:
+            if not self.readonly:
+                self._connection.commit()
+        if self.readonly:
+            return self._cursor.fetchall()
+
+    def execute_batch(self, query, vals, retry_counter=0):
+        try:
+            psycopg2.extras.execute_batch(self._cursor, query, vals)
+            retry_counter = 0
+        except (psycopg2.DatabaseError, psycopg2.OperationalError) as error:
+            if retry_counter >= LIMIT_RETRIES:
+                raise error
+            retry_counter += 1
+            print("got error {}. retrying {}".format(str(error).strip(), retry_counter))
+            time.sleep(1)
+            self.reset()
+            self.execute_batch(query, vals, retry_counter)
+        except (Exception, psycopg2.Error) as error:
+            if not self.readonly:
+                self._connection.rollback()
+            raise error
+        else:
+            if not self.readonly:
+                self._connection.commit()
+        if self.readonly:
+            return self._cursor.fetchall()
+
+    def execute_values(self, query, vals, page_size=1000, retry_counter=0):
+        try:
+            psycopg2.extras.execute_values(
+                self._cursor, query, vals, page_size=page_size
+            )
+            retry_counter = 0
+        except (psycopg2.DatabaseError, psycopg2.OperationalError) as error:
+            if retry_counter >= LIMIT_RETRIES:
+                raise error
+            retry_counter += 1
+            print("got error {}. retrying {}".format(str(error).strip(), retry_counter))
+            time.sleep(1)
+            self.reset()
+            self.execute_values(query, vals, page_size, retry_counter)
+        except (Exception, psycopg2.Error) as error:
+            if not self.readonly:
+                self._connection.rollback()
+            raise error
+        if not self.readonly:
```

Comment:
this cannot be readonly this is why there is no related code here right?

Reply:
readonly checks are for commit and rollback. in our cases execute_values only write